File: windows/CursorTokenCore/CursorAuth.cs
using System.Text.Encodings.Web;
using System.Text.Json;

namespace CursorTokenCore;

public sealed record CursorInstall(string Name, string StateDb, string? LaunchPath);

public sealed record CursorAuthApplyResult(
    bool Ok,
    string Message,
    string? BackupPath = null,
    int KeysWritten = 0,
    string? InstallName = null,
    bool Relaunched = false);

/// <summary>
/// Surgical read/write of Cursor's <c>state.vscdb</c> auth keys.
/// Browser <c>web</c> cookies cannot sign the desktop app in; only a <c>session</c> JWT can.
/// </summary>
public static class CursorAuth
{
    public const string BackupSuffix = ".tray-backup";
    public const string WebTokenMessage =
        "浏览器 Cookie（JWT type=web）不能登录 Cursor 应用，写回去会把客户端登出。"
        + "请先在 Cursor 里登录该账号，再点「从 Cursor 导入」，之后才能写回切换。";
    public const string NotJwtMessage = "当前 Token 不是 Cursor 会话，无法写入客户端。请先从 Cursor 应用导入。";
    public const string ConfirmCloseMessage =
        "会先关闭 Cursor，再把当前账号写入客户端。请先保存未提交的改动。未保存的文件不会被强制结束。";

    public static readonly string[] OwnedKeys =
    [
        "cursorAuth/accessToken",
        "cursorAuth/refreshToken",
        "cursorAuth/cachedEmail",
        "cursorAuth/cachedSignUpType",
        "cursorAuth/cachedScopedProfile",
        "cursorAuth/stripeMembershipType",
        "cursorAuth/stripeSubscriptionStatus",
        "glass.lastSignedInAuthId",
    ];

    public static readonly string[] OptionalUpdateKeys = ["cursorAuth/cachedAccessToken"];

    public static bool CanWrite(string token) => Token.CanWriteToCursor(token);

    public static (Dictionary<string, string>? Values, string Error) BuildValues(
        string token,
        string? email = null,
        string? membershipType = null,
        string? displayName = null)
    {
        var jwt = Token.ExtractJwt(token);
        if (!Token.LooksLikeJwt(jwt))
            return (null, NotJwtMessage);
        if (Token.JwtType(jwt) == "web")
            return (null, WebTokenMessage);

        var values = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["cursorAuth/accessToken"] = jwt,
            ["cursorAuth/refreshToken"] = jwt,
            ["cursorAuth/cachedSignUpType"] = "Auth_0",
        };

        var sub = Token.JwtSubject(jwt).Trim();
        if (sub.Length == 0)
        {
            var uid = Token.AccountId(token);
            if (uid.Length > 0 && !uid.StartsWith("tok_", StringComparison.Ordinal))
                sub = "auth0|" + uid;
        }
        if (sub.Length > 0) values["glass.lastSignedInAuthId"] = sub;

        var mail = FirstNonEmpty(Token.JwtClaim(jwt, "email"), email);
        if (LooksLikeEmail(mail)) values["cursorAuth/cachedEmail"] = mail.Trim();

        var name = FirstNonEmpty(displayName, mail);
        if (!string.IsNullOrWhiteSpace(name))
            values["cursorAuth/cachedScopedProfile"] = JsonSerializer.Serialize(
                new Dictionary<string, string> { ["displayName"] = name.Trim() },
                new JsonSerializerOptions { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping });

        var (plan, status) = StripePlan(membershipType);
        if (plan.Length > 0)
        {
            values["cursorAuth/stripeMembershipType"] = plan;
            values["cursorAuth/stripeSubscriptionStatus"] = status;
        }
        return (values, "");
    }

    public static (string Membership, string Status) StripePlan(string? membershipType)
    {
        var raw = (membershipType ?? "").Trim();
        if (raw.Length == 0 || raw == "未知") return ("", "");
        var lower = raw.ToLowerInvariant();
        var key = lower switch
        {
            "pro" or "pro+" or "pro_plus" => lower == "pro+" ? "pro_plus" : lower,
            "ultra" => "ultra",
            "free" or "hobby" or "unpaid" => "free",
            "enterprise" or "enterprise_trial" => lower,
            "team" or "teams" => "team",
            "business" => "business",
            _ => "",
        };
        if (key.Length == 0)
        {
            if (raw.Equals("Pro", StringComparison.OrdinalIgnoreCase)) key = "pro";
            else if (raw.Equals("Pro+", StringComparison.OrdinalIgnoreCase)) key = "pro_plus";
            else if (raw.Equals("Ultra", StringComparison.OrdinalIgnoreCase)) key = "ultra";
            else if (raw.Equals("Free", StringComparison.OrdinalIgnoreCase) || raw.Equals("Hobby", StringComparison.OrdinalIgnoreCase)) key = "free";
            else if (raw.Equals("Enterprise", StringComparison.OrdinalIgnoreCase)) key = "enterprise";
            else if (raw.Equals("Team", StringComparison.OrdinalIgnoreCase)) key = "team";
            else if (raw.Equals("Business", StringComparison.OrdinalIgnoreCase)) key = "business";
            else key = lower;
        }
        var unpaid = key is "free" or "hobby" or "unpaid";
        return (unpaid ? "free" : key, unpaid ? "unpaid" : "active");
    }

    public static Dictionary<string, string> ReadValues(string dbPath, IEnumerable<string>? keys = null)
    {
        var found = new Dictionary<string, string>(StringComparer.Ordinal);
        if (!File.Exists(dbPath)) return found;
        var want = (keys ?? OwnedKeys.Concat(OptionalUpdateKeys)).ToHashSet(StringComparer.Ordinal);
        try
        {
            using var conn = Open(dbPath, readOnly: true);
            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT key, value FROM ItemTable";
            using var r = cmd.ExecuteReader();
            while (r.Read())
            {
                var key = r.IsDBNull(0) ? "" : r.GetString(0);
                if (!want.Contains(key)) continue;
                var value = r.IsDBNull(1) ? "" : r.GetValue(1)?.ToString() ?? "";
                if (key.Length > 0) found[key] = value;
            }
        }
        catch { }
        return found;
    }

    public static CursorAuthApplyResult WriteValues(string dbPath, IReadOnlyDictionary<string, string> values, bool backup = true)
    {
        if (string.IsNullOrWhiteSpace(dbPath) || !File.Exists(dbPath))
            return new CursorAuthApplyResult(false, "未找到 Cursor 状态库，请先至少启动并登录过一次 Cursor。");
        if (values.Count == 0)
            return new CursorAuthApplyResult(false, "没有可写入的登录态。");

        string? backupPath = null;
        if (backup)
        {
            try { backupPath = Backup(dbPath); }
            catch (Exception ex)
            {
                return new CursorAuthApplyResult(false, "备份 Cursor 状态库失败，已中止写入：" + ex.Message);
            }
        }

        try
        {
            using var conn = Open(dbPath, readOnly: false);
            using var tx = conn.BeginTransaction();
            var existing = new HashSet<string>(StringComparer.Ordinal);
            using (var list = conn.CreateCommand())
            {
                list.Transaction = tx;
                list.CommandText = "SELECT key FROM ItemTable";
                using var r = list.ExecuteReader();
                while (r.Read())
                {
                    var key = r.IsDBNull(0) ? "" : r.GetString(0);
                    if (key.Length > 0) existing.Add(key);
                }
            }

            var written = 0;
            using var cmd = conn.CreateCommand();
            cmd.Transaction = tx;
            cmd.CommandText = "INSERT OR REPLACE INTO ItemTable(key, value) VALUES ($k, $v)";
            var pk = cmd.Parameters.Add("$k", Microsoft.Data.Sqlite.SqliteType.Text);
            var pv = cmd.Parameters.Add("$v", Microsoft.Data.Sqlite.SqliteType.Text);
            foreach (var key in OwnedKeys)
            {
                if (!values.TryGetValue(key, out var value) || string.IsNullOrEmpty(value)) continue;
                pk.Value = key;
                pv.Value = value;
                cmd.ExecuteNonQuery();
                written++;
            }
            foreach (var key in OptionalUpdateKeys)
            {
                if (!existing.Contains(key)) continue;
                if (!values.TryGetValue("cursorAuth/accessToken", out var jwt) || jwt.Length == 0) continue;
                pk.Value = key;
                pv.Value = jwt;
                cmd.ExecuteNonQuery();
                written++;
            }
            tx.Commit();
            return new CursorAuthApplyResult(true, $"已写入 Cursor（{written} 项）", backupPath, written);
        }
        catch (Exception ex)
        {
            return new CursorAuthApplyResult(false, "写入 Cursor 状态库失败：" + ex.Message, backupPath);
        }
    }

    public static string Backup(string dbPath)
    {
        var dest = dbPath + BackupSuffix;
        File.Copy(dbPath, dest, true);
        foreach (var suffix in new[] { "-wal", "-shm" })
        {
            var side = dbPath + suffix;
            if (File.Exists(side))
            {
                try { File.Copy(side, dest + suffix, true); } catch { }
            }
        }
        return dest;
    }

    public static IEnumerable<CursorInstall> FindInstalls(string? appData = null, string? localAppData = null, string? home = null)
    {
        var roaming = appData
            ?? Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        if (string.IsNullOrEmpty(roaming) && !string.IsNullOrEmpty(home))
            roaming = Path.Combine(home, "AppData", "Roaming");
        var local = localAppData
            ?? Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        if (string.IsNullOrEmpty(local) && !string.IsNullOrEmpty(home))
            local = Path.Combine(home, "AppData", "Local");

        foreach (var (name, folder) in new[]
        {
            ("Cursor", "Cursor"),
            ("Cursor Nightly", "Cursor Nightly"),
            ("Cursor Insiders", "Cursor - Insiders"),
        })
        {
            var db = Path.Combine(roaming, folder, "User", "globalStorage", "state.vscdb");
            if (!File.Exists(db)) continue;
            yield return new CursorInstall(name, db, FindLaunchPath(name, local));
        }
    }

    public static CursorInstall? ResolveTarget(IEnumerable<CursorInstall>? installs = null)
    {
        var list = (installs ?? FindInstalls()).ToList();
        if (list.Count == 0) return null;
        foreach (var inst in list)
        {
            if (IsRunning(inst)) return inst;
        }
        return list[0];
    }

    public static bool IsRunning(CursorInstall install)
    {
        try
        {
            foreach (var proc in System.Diagnostics.Process.GetProcesses())
            {
                try
                {
                    if (!IsCursorProcess(proc, install)) continue;
                    if (!proc.HasExited) return true;
                }
                catch { }
            }
        }
        catch { }
        return false;
    }

    public static bool RequestClose(CursorInstall install)
    {
        var any = false;
        try
        {
            foreach (var proc in System.Diagnostics.Process.GetProcesses())
            {
                try
                {
                    if (!IsCursorProcess(proc, install)) continue;
                    if (proc.HasExited) continue;
                    any = true;
                    try { proc.CloseMainWindow(); } catch { }
                }
                catch { }
            }
        }
        catch { }
        return any;
    }

    public static bool WaitUntilGone(CursorInstall install, TimeSpan timeout)
    {
        var deadline = DateTime.UtcNow + timeout;
        while (DateTime.UtcNow < deadline)
        {
            if (!IsRunning(install)) return true;
            Thread.Sleep(250);
        }
        return !IsRunning(install);
    }

    public static bool Launch(CursorInstall install)
    {
        var path = install.LaunchPath;
        if (string.IsNullOrEmpty(path) || !File.Exists(path))
            path = FindLaunchPath(install.Name, Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData));
        if (string.IsNullOrEmpty(path) || !File.Exists(path)) return false;
        try
        {
            System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo(path) { UseShellExecute = true });
            return true;
        }
        catch { return false; }
    }

    public static CursorAuthApplyResult Apply(
        string token,
        string? email = null,
        string? membershipType = null,
        string? displayName = null,
        bool closeIfRunning = true,
        bool relaunch = true,
        TimeSpan? wait = null,
        IEnumerable<CursorInstall>? installs = null,
        Func<CursorInstall, bool>? isRunning = null,
        Func<CursorInstall, bool>? requestClose = null,
        Func<CursorInstall, bool>? waitGone = null,
        Func<CursorInstall, bool>? launch = null,
        Func<string, IReadOnlyDictionary<string, string>, bool, CursorAuthApplyResult>? write = null)
    {
        var built = BuildValues(token, email, membershipType, displayName);
        if (built.Values is null) return new CursorAuthApplyResult(false, built.Error);

        var target = ResolveTarget(installs);
        if (target is null)
            return new CursorAuthApplyResult(false, "未找到 Cursor 状态库，请先至少启动并登录过一次 Cursor。");

        var runningFn = isRunning ?? IsRunning;
        var closeFn = requestClose ?? RequestClose;
        var goneFn = waitGone ?? (inst => WaitUntilGone(inst, wait ?? TimeSpan.FromSeconds(10)));
        var launchFn = launch ?? Launch;
        var writeFn = write ?? WriteValues;

        if (runningFn(target))
        {
            if (!closeIfRunning)
                return new CursorAuthApplyResult(false, "Cursor 正在运行。请先关闭后再写入，以免冲掉未保存的文件。", InstallName: target.Name);
            closeFn(target);
            if (!goneFn(target))
                return new CursorAuthApplyResult(false,
                    "Cursor 没有退出（可能有未保存的改动）。请先手动关闭后再试，不会强制结束。",
                    InstallName: target.Name);
        }

        var written = writeFn(target.StateDb, built.Values, true);
        if (!written.Ok) return written with { InstallName = target.Name };

        var started = false;
        if (relaunch) started = launchFn(target);
        var msg = written.Message + " · " + target.Name;
        if (relaunch)
            msg += started ? "，已重新打开。" : "。请手动打开 Cursor。";
        else
            msg += "。请重新打开 Cursor 后生效。";
        return written with { Message = msg, InstallName = target.Name, Relaunched = started };
    }

    static Microsoft.Data.Sqlite.SqliteConnection Open(string dbPath, bool readOnly)
    {
        var conn = new Microsoft.Data.Sqlite.SqliteConnection(new Microsoft.Data.Sqlite.SqliteConnectionStringBuilder
        {
            DataSource = dbPath,
            Mode = readOnly
                ? Microsoft.Data.Sqlite.SqliteOpenMode.ReadOnly
                : Microsoft.Data.Sqlite.SqliteOpenMode.ReadWrite,
        }.ToString());
        conn.Open();
        using var pragma = conn.CreateCommand();
        pragma.CommandText = "PRAGMA busy_timeout = 5000";
        pragma.ExecuteNonQuery();
        return conn;
    }

    static bool IsCursorProcess(System.Diagnostics.Process proc, CursorInstall install)
    {
        var name = proc.ProcessName;
        if (!name.Contains("Cursor", StringComparison.OrdinalIgnoreCase)) return false;
        string? file = null;
        try { file = proc.MainModule?.FileName; } catch { }
        if (string.IsNullOrEmpty(file))
        {
            // If we cannot see the path, only match the default Cursor install by process name.
            return install.Name == "Cursor" && name.Equals("Cursor", StringComparison.OrdinalIgnoreCase);
        }
        var hay = file.Replace('/', Path.DirectorySeparatorChar);
        if (install.Name.Contains("Nightly", StringComparison.OrdinalIgnoreCase))
            return hay.Contains("Cursor Nightly", StringComparison.OrdinalIgnoreCase);
        if (install.Name.Contains("Insiders", StringComparison.OrdinalIgnoreCase))
            return hay.Contains("Cursor - Insiders", StringComparison.OrdinalIgnoreCase)
                || hay.Contains("Cursor Insiders", StringComparison.OrdinalIgnoreCase);
        return hay.Contains($"{Path.DirectorySeparatorChar}Cursor{Path.DirectorySeparatorChar}", StringComparison.OrdinalIgnoreCase)
            || hay.EndsWith($"{Path.DirectorySeparatorChar}Cursor.exe", StringComparison.OrdinalIgnoreCase);
    }

    static string? FindLaunchPath(string name, string localAppData)
    {
        if (string.IsNullOrEmpty(localAppData)) return null;
        var folders = name.Contains("Nightly", StringComparison.OrdinalIgnoreCase)
            ? new[] { "cursor nightly", "Cursor Nightly" }
            : name.Contains("Insiders", StringComparison.OrdinalIgnoreCase)
                ? new[] { "cursor insiders", "Cursor Insiders", "cursor-insiders" }
                : new[] { "cursor", "Cursor" };
        foreach (var folder in folders)
        {
            var candidate = Path.Combine(localAppData, "Programs", folder, "Cursor.exe");
            if (File.Exists(candidate)) return candidate;
        }
        return null;
    }

    static string FirstNonEmpty(params string?[] values)
    {
        foreach (var v in values)
        {
            if (!string.IsNullOrWhiteSpace(v)) return v!;
        }
        return "";
    }

    public static bool LooksLikeEmail(string? value)
    {
        var text = (value ?? "").Trim();
        var at = text.IndexOf('@');
        return at > 0 && at < text.Length - 1 && !text.Contains(' ');
    }
}

File: macos/Sources/CursorTokenCore/CursorAuth.swift
import Foundation

#if canImport(SQLite3)
import SQLite3
#endif

public struct CursorInstall: Equatable, Sendable {
    public var name: String
    public var stateDb: URL
    public var launchPath: URL?

    public init(name: String, stateDb: URL, launchPath: URL? = nil) {
        self.name = name
        self.stateDb = stateDb
        self.launchPath = launchPath
    }
}

public struct CursorAuthApplyResult: Equatable, Sendable {
    public var ok: Bool
    public var message: String
    public var backupPath: String?
    public var keysWritten: Int
    public var installName: String?
    public var relaunched: Bool

    public init(
        ok: Bool,
        message: String,
        backupPath: String? = nil,
        keysWritten: Int = 0,
        installName: String? = nil,
        relaunched: Bool = false
    ) {
        self.ok = ok
        self.message = message
        self.backupPath = backupPath
        self.keysWritten = keysWritten
        self.installName = installName
        self.relaunched = relaunched
    }
}

public enum CursorAuth {
    public static let backupSuffix = ".tray-backup"
    public static let webTokenMessage =
        "浏览器 Cookie（JWT type=web）不能登录 Cursor 应用，写回去会把客户端登出。"
        + "请先在 Cursor 里登录该账号，再点「从 Cursor 导入」，之后才能写回切换。"
    public static let notJwtMessage = "当前 Token 不是 Cursor 会话，无法写入客户端。请先从 Cursor 应用导入。"
    public static let confirmCloseMessage =
        "会先关闭 Cursor，再把当前账号写入客户端。请先保存未提交的改动。未保存的文件不会被强制结束。"

    public static let ownedKeys = [
        "cursorAuth/accessToken",
        "cursorAuth/refreshToken",
        "cursorAuth/cachedEmail",
        "cursorAuth/cachedSignUpType",
        "cursorAuth/cachedScopedProfile",
        "cursorAuth/stripeMembershipType",
        "cursorAuth/stripeSubscriptionStatus",
        "glass.lastSignedInAuthId",
    ]
    public static let optionalUpdateKeys = ["cursorAuth/cachedAccessToken"]

    public static func canWrite(_ token: String) -> Bool { Token.canWriteToCursor(token) }

    public static func buildValues(
        token: String,
        email: String? = nil,
        membershipType: String? = nil,
        displayName: String? = nil
    ) -> (values: [String: String]?, error: String) {
        let jwt = Token.extractJWT(token)
        if !Token.looksLikeJWT(jwt) {
            return (nil, notJwtMessage)
        }
        if Token.jwtType(jwt) == "web" {
            return (nil, webTokenMessage)
        }
        var values: [String: String] = [
            "cursorAuth/accessToken": jwt,
            "cursorAuth/refreshToken": jwt,
            "cursorAuth/cachedSignUpType": "Auth_0",
        ]
        var sub = Token.jwtSubject(jwt).trimmingCharacters(in: .whitespaces)
        if sub.isEmpty {
            let uid = Token.accountId(from: token)
            if !uid.isEmpty && !uid.hasPrefix("tok_") {
                sub = "auth0|\(uid)"
            }
        }
        if !sub.isEmpty { values["glass.lastSignedInAuthId"] = sub }

        let mail = firstNonEmpty(Token.jwtClaim(jwt, "email"), email)
        if looksLikeEmail(mail) { values["cursorAuth/cachedEmail"] = mail.trimmingCharacters(in: .whitespaces) }

        let name = firstNonEmpty(displayName, mail)
        if !name.trimmingCharacters(in: .whitespaces).isEmpty {
            let profile = ["displayName": name.trimmingCharacters(in: .whitespaces)]
            if let data = try? JSONSerialization.data(withJSONObject: profile),
               let json = String(data: data, encoding: .utf8)
            {
                values["cursorAuth/cachedScopedProfile"] = json
            }
        }

        let plan = stripePlan(membershipType)
        if !plan.membership.isEmpty {
            values["cursorAuth/stripeMembershipType"] = plan.membership
            values["cursorAuth/stripeSubscriptionStatus"] = plan.status
        }
        return (values, "")
    }

    public static func stripePlan(_ membershipType: String?) -> (membership: String, status: String) {
        let raw = (membershipType ?? "").trimmingCharacters(in: .whitespaces)
        if raw.isEmpty || raw == "未知" { return ("", "") }
        let lower = raw.lowercased()
        var key = ""
        switch lower {
        case "pro": key = "pro"
        case "pro+", "pro_plus": key = "pro_plus"
        case "ultra": key = "ultra"
        case "free", "hobby", "unpaid": key = "free"
        case "enterprise", "enterprise_trial": key = lower
        case "team", "teams": key = "team"
        case "business": key = "business"
        default:
            switch raw.lowercased() {
            case "pro": key = "pro"
            case "pro+": key = "pro_plus"
            case "ultra": key = "ultra"
            case "free", "hobby": key = "free"
            case "enterprise": key = "enterprise"
            case "team": key = "team"
            case "business": key = "business"
            default: key = lower
            }
        }
        let unpaid = ["free", "hobby", "unpaid"].contains(key)
        return (unpaid ? "free" : key, unpaid ? "unpaid" : "active")
    }

    public static func readValues(dbPath: URL, keys: [String]? = nil) -> [String: String] {
        var found: [String: String] = [:]
        guard FileManager.default.fileExists(atPath: dbPath.path) else { return found }
        let want = Set(keys ?? ownedKeys + optionalUpdateKeys)
        #if canImport(SQLite3)
        var db: OpaquePointer?
        guard sqlite3_open_v2(dbPath.path, &db, SQLITE_OPEN_READONLY, nil) == SQLITE_OK, let db else { return found }
        defer { sqlite3_close(db) }
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, "SELECT key, value FROM ItemTable", -1, &stmt, nil) == SQLITE_OK, let stmt else {
            return found
        }
        defer { sqlite3_finalize(stmt) }
        while sqlite3_step(stmt) == SQLITE_ROW {
            let key = sqlite3_column_text(stmt, 0).map { String(cString: $0) } ?? ""
            guard want.contains(key) else { continue }
            let value = sqlite3_column_text(stmt, 1).map { String(cString: $0) } ?? ""
            if !key.isEmpty { found[key] = value }
        }
        #endif
        return found
    }

    public static func writeValues(dbPath: URL, values: [String: String], backup: Bool = true) -> CursorAuthApplyResult {
        guard FileManager.default.fileExists(atPath: dbPath.path) else {
            return CursorAuthApplyResult(ok: false, message: "未找到 Cursor 状态库，请先至少启动并登录过一次 Cursor。")
        }
        guard !values.isEmpty else {
            return CursorAuthApplyResult(ok: false, message: "没有可写入的登录态。")
        }
        var backupPath: String?
        if backup {
            do { backupPath = try Self.backup(dbPath: dbPath) }
            catch {
                return CursorAuthApplyResult(ok: false, message: "备份 Cursor 状态库失败，已中止写入：\(error.localizedDescription)")
            }
        }
        #if canImport(SQLite3)
        var db: OpaquePointer?
        guard sqlite3_open_v2(dbPath.path, &db, SQLITE_OPEN_READWRITE, nil) == SQLITE_OK, let db else {
            return CursorAuthApplyResult(ok: false, message: "无法打开 Cursor 状态库。", backupPath: backupPath)
        }
        defer { sqlite3_close(db) }
        sqlite3_busy_timeout(db, 5000)
        guard sqlite3_exec(db, "BEGIN IMMEDIATE", nil, nil, nil) == SQLITE_OK else {
            return CursorAuthApplyResult(ok: false, message: "无法开始写入 Cursor 状态库。", backupPath: backupPath)
        }
        var existing = Set<String>()
        var list: OpaquePointer?
        if sqlite3_prepare_v2(db, "SELECT key FROM ItemTable", -1, &list, nil) == SQLITE_OK, let list {
            defer { sqlite3_finalize(list) }
            while sqlite3_step(list) == SQLITE_ROW {
                if let c = sqlite3_column_text(list, 0) { existing.insert(String(cString: c)) }
            }
        }
        let sqliteTransient = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, "INSERT OR REPLACE INTO ItemTable(key, value) VALUES (?, ?)", -1, &stmt, nil) == SQLITE_OK, let stmt else {
            sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
            return CursorAuthApplyResult(ok: false, message: "无法准备写入 Cursor 状态库。", backupPath: backupPath)
        }
        defer { sqlite3_finalize(stmt) }
        var written = 0
        func bind(_ key: String, _ value: String) -> Bool {
            sqlite3_reset(stmt)
            sqlite3_clear_bindings(stmt)
            key.withCString { kptr in
                _ = sqlite3_bind_text(stmt, 1, kptr, -1, sqliteTransient)
            }
            value.withCString { vptr in
                _ = sqlite3_bind_text(stmt, 2, vptr, -1, sqliteTransient)
            }
            return sqlite3_step(stmt) == SQLITE_DONE
        }
        for key in ownedKeys {
            guard let value = values[key], !value.isEmpty else { continue }
            if !bind(key, value) {
                sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
                return CursorAuthApplyResult(ok: false, message: "写入 Cursor 状态库失败。", backupPath: backupPath)
            }
            written += 1
        }
        if let jwt = values["cursorAuth/accessToken"], !jwt.isEmpty {
            for key in optionalUpdateKeys where existing.contains(key) {
                if bind(key, jwt) { written += 1 }
            }
        }
        if sqlite3_exec(db, "COMMIT", nil, nil, nil) != SQLITE_OK {
            sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
            return CursorAuthApplyResult(ok: false, message: "提交 Cursor 状态库失败。", backupPath: backupPath)
        }
        return CursorAuthApplyResult(ok: true, message: "已写入 Cursor（\(written) 项）", backupPath: backupPath, keysWritten: written)
        #else
        return CursorAuthApplyResult(ok: false, message: "当前环境无法写入 SQLite。", backupPath: backupPath)
        #endif
    }

    public static func backup(dbPath: URL) throws -> String {
        let dest = URL(fileURLWithPath: dbPath.path + backupSuffix)
        let fm = FileManager.default
        if fm.fileExists(atPath: dest.path) { try fm.removeItem(at: dest) }
        try fm.copyItem(at: dbPath, to: dest)
        for suffix in ["-wal", "-shm"] {
            let side = URL(fileURLWithPath: dbPath.path + suffix)
            if fm.fileExists(atPath: side.path) {
                let sideDest = URL(fileURLWithPath: dest.path + suffix)
                try? fm.removeItem(at: sideDest)
                try? fm.copyItem(at: side, to: sideDest)
            }
        }
        return dest.path
    }

    public static func findInstalls(home: URL = FileManager.default.homeDirectoryForCurrentUser) -> [CursorInstall] {
        #if os(macOS)
        let support = home.appendingPathComponent("Library/Application Support")
        let apps = [
            ("Cursor", "Cursor", "/Applications/Cursor.app"),
            ("Cursor Nightly", "Cursor Nightly", "/Applications/Cursor Nightly.app"),
            ("Cursor Insiders", "Cursor - Insiders", "/Applications/Cursor - Insiders.app"),
        ]
        var found: [CursorInstall] = []
        for (name, folder, launch) in apps {
            let db = support.appendingPathComponent("\(folder)/User/globalStorage/state.vscdb")
            guard FileManager.default.fileExists(atPath: db.path) else { continue }
            let app = URL(fileURLWithPath: launch)
            found.append(CursorInstall(
                name: name,
                stateDb: db,
                launchPath: FileManager.default.fileExists(atPath: app.path) ? app : nil
            ))
        }
        return found
        #else
        _ = home
        return []
        #endif
    }

    public static func resolveTarget(_ installs: [CursorInstall]? = nil) -> CursorInstall? {
        let list = installs ?? findInstalls()
        if list.isEmpty { return nil }
        for inst in list where isRunning(inst) { return inst }
        return list[0]
    }

    public static func isRunning(_ install: CursorInstall) -> Bool {
        !runningPids(install).isEmpty
    }

    public static func requestClose(_ install: CursorInstall) -> Bool {
        let appName = install.name
        _ = run("/usr/bin/osascript", arguments: ["-e", "tell application \"\(appName)\" to quit"])
        let pids = runningPids(install)
        if pids.isEmpty { return true }
        for pid in pids {
            _ = run("/bin/kill", arguments: ["-TERM", String(pid)])
        }
        return true
    }

    public static func waitUntilGone(_ install: CursorInstall, timeout: TimeInterval = 10) -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            if !isRunning(install) { return true }
            Thread.sleep(forTimeInterval: 0.25)
        }
        return !isRunning(install)
    }

    public static func launch(_ install: CursorInstall) -> Bool {
        if let url = install.launchPath, FileManager.default.fileExists(atPath: url.path) {
            return run("/usr/bin/open", arguments: [url.path]).0 == 0
        }
        return run("/usr/bin/open", arguments: ["-a", install.name]).0 == 0
    }

    public static func apply(
        token: String,
        email: String? = nil,
        membershipType: String? = nil,
        displayName: String? = nil,
        closeIfRunning: Bool = true,
        relaunch: Bool = true,
        wait: TimeInterval = 10,
        installs: [CursorInstall]? = nil,
        isRunning: ((CursorInstall) -> Bool)? = nil,
        requestClose: ((CursorInstall) -> Bool)? = nil,
        waitGone: ((CursorInstall) -> Bool)? = nil,
        launch: ((CursorInstall) -> Bool)? = nil,
        write: ((URL, [String: String], Bool) -> CursorAuthApplyResult)? = nil
    ) -> CursorAuthApplyResult {
        let built = buildValues(token: token, email: email, membershipType: membershipType, displayName: displayName)
        guard let values = built.values else {
            return CursorAuthApplyResult(ok: false, message: built.error)
        }
        guard let target = resolveTarget(installs) else {
            return CursorAuthApplyResult(ok: false, message: "未找到 Cursor 状态库，请先至少启动并登录过一次 Cursor。")
        }
        let runningFn = isRunning ?? Self.isRunning
        let closeFn = requestClose ?? Self.requestClose
        let goneFn = waitGone ?? { Self.waitUntilGone($0, timeout: wait) }
        let launchFn = launch ?? Self.launch
        let writeFn = write ?? { Self.writeValues(dbPath: $0, values: $1, backup: $2) }

        if runningFn(target) {
            if !closeIfRunning {
                return CursorAuthApplyResult(ok: false, message: "Cursor 正在运行。请先关闭后再写入，以免冲掉未保存的文件。", installName: target.name)
            }
            _ = closeFn(target)
            if !goneFn(target) {
                return CursorAuthApplyResult(
                    ok: false,
                    message: "Cursor 没有退出（可能有未保存的改动）。请先手动关闭后再试，不会强制结束。",
                    installName: target.name
                )
            }
        }
        var written = writeFn(target.stateDb, values, true)
        if !written.ok {
            written.installName = target.name
            return written
        }
        var started = false
        if relaunch { started = launchFn(target) }
        var msg = written.message + " · " + target.name
        if relaunch {
            msg += started ? "，已重新打开。" : "。请手动打开 Cursor。"
        } else {
            msg += "。请重新打开 Cursor 后生效。"
        }
        written.message = msg
        written.installName = target.name
        written.relaunched = started
        return written
    }

    public static func looksLikeEmail(_ value: String?) -> Bool {
        let text = (value ?? "").trimmingCharacters(in: .whitespaces)
        guard let at = text.firstIndex(of: "@") else { return false }
        return at > text.startIndex && text.distance(from: at, to: text.endIndex) > 1 && !text.contains(" ")
    }

    static func runningPids(_ install: CursorInstall) -> [Int32] {
        let needle: String
        if install.name.contains("Nightly") {
            needle = "Cursor Nightly.app/Contents/MacOS"
        } else if install.name.contains("Insiders") {
            needle = "Cursor - Insiders.app/Contents/MacOS"
        } else {
            needle = "Cursor.app/Contents/MacOS"
        }
        let (status, output) = run("/usr/bin/pgrep", arguments: ["-f", needle])
        if status != 0 { return [] }
        return output.split(whereSeparator: \.isNewline).compactMap { Int32($0.trimmingCharacters(in: .whitespaces)) }
    }

    static func run(_ launchPath: String, arguments: [String]) -> (Int32, String) {
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: launchPath)
        proc.arguments = arguments
        let pipe = Pipe()
        proc.standardOutput = pipe
        proc.standardError = Pipe()
        do {
            try proc.run()
            proc.waitUntilExit()
        } catch {
            return (1, "")
        }
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        let text = String(data: data, encoding: .utf8) ?? ""
        return (proc.terminationStatus, text)
    }

    static func firstNonEmpty(_ values: String?...) -> String {
        for value in values {
            if let value, !value.trimmingCharacters(in: .whitespaces).isEmpty { return value }
        }
        return ""
    }
}
